File: Tests/Full/C#/CameraTests.cs
using System;
using System.Collections.Generic;
using System.Text;
using SwinGame;
using System.Drawing;

namespace Tests
{
    class CameraTests : IGameTestLoader
    {
        #region IGameTestLoader Members

        public void AddTo(List<TestSuite> list)
        {
            TestSuite result = new TestSuite("Camera Tests");
            result.Add(new TestCamera());
            list.Add(result);
        }

        #endregion

        private class TestCamera : TestSet
        {
            private readonly static string METHS =
                "All camera routines";

            private readonly static string INST =
                "Use the arrow keys to move" + Environment.NewLine +
                "the ship." + Environment.NewLine +
                "I: Follow the ship" + Environment.NewLine +
                "U: Unfollow" + Environment.NewLine +
                "W: Move up the visual area" + Environment.NewLine +
                "S: Move down the visual area" + Environment.NewLine +
                "A: Move the visual area to" + Environment.NewLine +
                "the left" + Environment.NewLine +
                "D: Move the visual area to" + Environment.NewLine +
                "the right";

			private SwinGame.Bitmap bgImage = GameResources.GameImage("Sea");
			private Sprite shipSprite = SwinGame.Graphics.CreateSprite(GameResources.GameImage("Ship"), 3, 2, 40, 43);
			private bool follow = false;

            public TestCamera() : base(METHS, INST) 
            {
                shipSprite.X = 100; 
                shipSprite.Y = 100;
            }


            protected override void ToRun(System.Drawing.Rectangle drawIn)
            {
                if (Input.IsKeyPressed(Keys.VK_RIGHT))  shipSprite.X = shipSprite.X + 4;
		        if (Input.IsKeyPressed(Keys.VK_DOWN))  shipSprite.Y = shipSprite.Y + 4;
		        if (Input.IsKeyPressed(Keys.VK_UP))  shipSprite.Y = shipSprite.Y - 4;
		        if (Input.IsKeyPressed(Keys.VK_LEFT))  shipSprite.X = shipSprite.X - 4;
		        if (Input.IsKeyPressed(Keys.VK_U))  follow = false;
		        if (Input.IsKeyPressed(Keys.VK_I))  follow = true;
		        if (Input.IsKeyPressed(Keys.VK_A))  Camera.MoveVisualArea(Physics.CreateVector(-20, 0));
                if (Input.IsKeyPressed(Keys.VK_D)) Camera.MoveVisualArea(Physics.CreateVector(20, 0));
                if (Input.IsKeyPressed(Keys.VK_W)) Camera.MoveVisualArea(0, -20);
                if (Input.IsKeyPressed(Keys.VK_S)) Camera.MoveVisualArea(0, 20);		
		        if (follow) Camera.FollowSprite(shipSprite, (int)(400 - drawIn.X - drawIn.Width / 2), -50);
        		
		        SwinGame.Graphics.DrawBitmap(bgImage, 0, 0);
                Text.DrawText("0, 0", Color.Red, GameResources.GameFont("Courier"), 0, 0);
                Text.DrawTextOnScreen("Ship Position: " + Convert.ToString(Camera.GameX(Camera.ScreenX(shipSprite.X))) + ", " + Convert.ToString(Camera.GameY(Camera.ScreenY(shipSprite.Y))), Color.Red, GameResources.GameFont("Courier"), (int)(drawIn.X), (int)(drawIn.Y));
                Text.DrawTextOnScreen("Ship Position On Screen: " + Convert.ToString(Camera.ScreenX(shipSprite.X)) + ", " + Convert.ToString(Camera.ScreenY(shipSprite.Y)), Color.Red, GameResources.GameFont("Courier"), (int)(drawIn.X), (int)(drawIn.Y) + 15);
                Text.DrawTextOnScreen("Camera Offset: " + Convert.ToString(Camera.XOffset) + ", " + Convert.ToString(Camera.YOffset), Color.Red, GameResources.GameFont("Courier"), (int)(drawIn.X), (int)(drawIn.Y) + 30);


                SwinGame.Graphics.DrawSprite(shipSprite);
                SwinGame.Graphics.UpdateSprite(shipSprite);
            }
        }
    }
}

File: Tests/Full/Pascal/src/CameraTests.pas
unit CameraTests;

interface
	uses TestFramework;

	procedure AddCameraSuite(var suites: TestSuites); 

implementation
	uses GameResources, sgCamera, sgImages, sgInput, sgGraphics, sgTypes,sgGeometry,sgText,sgPhysics, SysUtils, sgSprites, sgAnimations;
	
	var
		bgImage: Bitmap;
		shipSprite, ship2: Sprite;
		follow: Boolean;
	
	procedure TestCamera(const drawIn: Rectangle);
	var
	  mvmt: Vector;
	begin
	  mvmt := VectorTo(0,0);
	  
		if KeyDown(VK_RIGHT) then mvmt.x := mvmt.x + 4;
		if KeyDown(VK_DOWN) then mvmt.y := mvmt.y + 4;
		if KeyDown(VK_UP) then mvmt.y := mvmt.y - 4;
		if KeyDown(VK_LEFT) then mvmt.x := mvmt.x - 4;
		if KeyDown(VK_U) then follow := false;
		if KeyDown(VK_I) then follow := true;
		if KeyDown(VK_A) then MoveCameraBy(VectorTo(-20, 0));
		if KeyDown(VK_D) then MoveCameraBy(VectorTo(20, 0));
		if KeyDown(VK_W) then MoveCameraBy(0, -20);
		if KeyDown(VK_S) then MoveCameraBy(0, 20);
		  
	  MoveSprite(shipSprite, mvmt);	
		  
		if KeyDown(VK_COMMA) then SpriteSetRotation(shipSprite, SpriteRotation(shipSprite) + 5);
		if KeyDown(VK_PERIOD) then SpriteSetRotation(shipSprite, SpriteRotation(shipSprite) - 5);
		if follow then CenterCameraOn(shipSprite, Round(400 - drawIn.x - drawIn.width / 2), -50);
		
		DrawBitmap(bgImage, 0, 0);
		DrawText('0, 0', ColorRed, FontNamed('Courier'), 0, 0);
		DrawTextOnScreen('Ship Position: ' + FloatToStr(ToWorldX(ToScreenX(SpriteX(shipSprite)))) + ', ' + FloatToStr(ToWorldY(ToScreenY(SpriteY(shipSprite)))), ColorRed, FontNamed('Courier'), Round(drawIn.x), Round(drawIn.y));
		DrawTextOnScreen('Ship Position On Screen: ' + FloatToStr(ToScreenX(SpriteX(shipSprite))) + ', ' + FloatToStr(ToScreenY(SpriteY(shipSprite))), ColorRed, FontNamed('Courier'), Round(drawIn.x), Round(drawIn.y) + 15);
		DrawTextOnScreen('Camera Offset: ' + FloatToStr(CameraX()) + ', ' + FloatToStr(CameraY()), ColorRed, FontNamed('Courier'), Round(drawIn.x), Round(drawIn.y) + 30);
		
		if SpriteCollision(shipSprite, ship2) then
		  DrawTextOnScreen('Collision', ColorWhite, FontNamed('Courier'), Round(drawIn.x), Round(drawIn.y) + 45);
		
		DrawSprite(ship2);
		DrawSprite(shipSprite);
		DrawBitmapOnScreen(shipSprite^.collisionBitmap, 100, 200);
		if assigned(shipSprite^.cacheImage) then
		begin
		  DrawBitmapOnScreen(shipSprite^.cacheImage, 100, 150);
		end;
		UpdateSprite(shipSprite);
	end;

	function GetCameraTests(): TestSuite;
	var
		i: Integer;
	begin
		result.Title := 'Camera Test';
		SetLength(result.Tests, 1);
		
		for i := 0 to High(result.Tests) do
		begin
			InitTest(result.Tests[i]);
		end;
		
		with result.Tests[0] do
		begin
			MethodBeingTested := 'All camera routines';
			Instructions := 'Use the arrow keys to move' + LineEnding + 'the ship.' + LineEnding + 'I: Follow the ship' + LineEnding + 'U: Unfollow' + LineEnding + 'W: Move up the visual area' + LineEnding + 'S: Move down the visual area'
							 + LineEnding + 'A: Move the visual area to' + LineEnding + 'the left' + LineEnding + 'D: Move the visual area to' + LineEnding + 'the right';
			bgImage := BitmapNamed('Sea');
			shipSprite := CreateSprite(BitmapNamed('Ship'), LoadAnimationScript('ship_anim.txt'));
			SpriteStartAnimation(shipSprite, 0);
			SpriteSetX(shipSprite, 100); 
			SpriteSetY(shipSprite, 100);
			follow := false;
			
			ship2 := CreateSprite(BitmapNamed('Ship'), LoadAnimationScript('ship_anim.txt'));
			SpriteStartAnimation(shipSprite, 0);
			SpriteSetX(shipSprite, 50); 
			SpriteSetY(shipSprite, 50);
			
			ToRun := @TestCamera;
		end;
	end;
	
	procedure AddCameraSuite(var suites: TestSuites);
	begin
		suites[High(suites)] := GetCameraTests();
	end;

end.